### PR TITLE
Removes JSON types and makes absence of (inherited) $type invalid

### LINKED ISSUE #120: Remove JSON derived types from the specification
https://design-tokens.github.io/community-group/format/#types

This part of the specification assumes that implementers will have an inherent understanding of certain types and how to handle them.

This implicit typing will lead to interop issues.
It's better to have exact definitions for each type supported in the schema.

An obvious example is `String`.
When converting a value to CSS it is ambiguous if it should become an ident token or a string token.

```json
{
  "$value": "center"
}
```

```css
.foo {
  content: "center";
}
```

or

```css
.foo {
  text-align: center;
}
```

You also can't do this in the tokens file :

```json
{
  "$value": "\"center\""
}
```

Because that dictates how strings work in any consuming context.

### PR DESCRIPTION
**Pull request by c1rrus**
Closes #120, #139

* Removes support for the JSON types (`string`, `boolean`, `null`, etc.)
    * Except for `number` which is now documented as one of the spec's types. This is because it was required by the `gradient` type.
* Changes the type resolution rules so that when there is no (inherited) `$type` present, a token must be considered invalid. Previously, it would fall back on to one of the JSON types.
* Fixes a bug (?) in the `typography` type, where `lineHeight` expected a `string` value.
    * I had to change this since `string` no longer exists. But also, it was never a useful type of value for that purpose. I've changed it to expect `number` values for now. We can always revise it if needed once we review and action the feedback provided in #102 
* Updates examples throughout the spec to ensure that they are valid now that the JSON types and associated type resolution fallback behaviour have been removed.

### CODE REVIEW

File: technical-reports/format/composite-types.md [hunk 1]
```diff
@@ -439,7 +440,7 @@ Represents a typographic style. The `$type` property MUST be set to the string `
         "fontSize": "{font.size.smallest}",
         "fontWeight": "{font.weight.normal}",
         "letterSpacing": "0px",
-        "lineHeight": "1"
+        "lineHeight": "24px"
```
Comment (romainmenke) [inline technical-reports/format/composite-types.md:443]: The type of `lineHeight` is ambiguous if both `1.2` and `"24px"` are allowed.

It is no longer possible to determine the value based on the current specification.
The specification only describes the `$type` property as a source for type information.

It is not possible to allow multiple types on the same field in a composite token.

**Review by romainmenke — CHANGES_REQUESTED** [review]

File: technical-reports/format/composite-types.md (hunk 1 above)
Comment (c1rrus) [reply technical-reports/format/composite-types.md:443]: I think it's a bit of a grey area. The spec just says that "names and types of sub-values are pre-defined". That doesn't necessarily _forbid_ a composite type for specifying a range of permitted types for one of its sub-values.

However, I agree that this probably needs to be made more explicit one way or the other. I only updated the `typography` type in this PR because `string` is now gone. To avoid this PR getting side-tracked on updates to the `typography` type (which is not this PR's purpose), I can perhaps change it to only accept `dimension` values for now. That's hopefully a more "minimal" change that achieves the original aim of not replacing the `string` type with something that's still valid.

Comment (romainmenke) [reply technical-reports/format/composite-types.md:443]: > I think it's a bit of a grey area. The spec just says that "names and types of sub-values are pre-defined". That doesn't necessarily forbid a composite type for specifying a range of permitted types for one of its sub-values.

This is definitely not a grey area :)
The specification doesn't define how a token parser must handle a composite token with fields that can have one of multiple types. Because it is not defined it can not be implemented. That it can not be implemented means that it isn't allowed/supported.

I think this issue is relevant : https://github.com/design-tokens/community-group/issues/199

-----

> I can perhaps change it to only accept dimension values for now.

Isn't `number` a better fit?
All the example previously were numbers but "embedded" in a JSON string.

I am unsure if this was something that leaked from CSS or not as it is typical for CSS that line height is a fraction of font size and therefor not a dimension.

Comment (c1rrus) [reply technical-reports/format/composite-types.md]: I've changed it to just `number` for now. Like you said, it's closer to what we had previously.

Comment (romainmenke) [reply technical-reports/format/composite-types.md]: Thank you for these changes 🙇

Comment (c1rrus) [reply technical-reports/format/composite-types.md]: Thanks for your review comments :-)

File: technical-reports/format/design-token.md [hunk 2]
```diff
@@ -7,19 +7,23 @@
...
 {
   "token name": {
-    "$value": "token value"
+    "$value": "#fff000",
+    "$type": "color"
   }
 }
 ```
 
+Note: The `$type` property has been added to ensure this example is valid. Please refer to the [type chapter](#type-0) for more details.
```
Comment (CITguy) [inline technical-reports/format/design-token.md:16]: Should the link be reworded to disambiguate between "5.4 Type", "6.1.2 Type", and "8 Types"?

e.g.,...

```markdown
... Please refer to the [design token type property chapter](#type-0) for more details.
```

File: technical-reports/format/design-token.md [hunk 3]
```diff
@@ -113,7 +125,7 @@ The `$type` property can be set on different levels:
 - at the group level
 - at the token level
 
-The `$type` property MUST be a plain JSON string, whose value is `string`, `number`, `boolean`, `object`, `array`, `null` or one of the values specified in respective [type chapters](#types). The value of `$type` is case-sensitive.
+The `$type` property MUST be a plain JSON string, whose value is one of the values specified in respective [type chapters](#types). The value of `$type` is case-sensitive.
```
Comment (CITguy) [inline technical-reports/format/design-token.md:128]: * I'd recommend removing the link to `#types` ("8 Types"), purely because local context would require linking to BOTH "8 Types" and "9 Composite types".  
* Additionally, I'd reword it as `... specified in respective type definitions.`

With the above changes, if § 8 and 9 were to be merged/reorganized in the future, then this content would still be correct and we could potentially reintroduce the link later.

```markdown
...values specified in respective type definitions. The value ...
```

File: technical-reports/format/design-token.md [hunk 4]
```diff
@@ -89,6 +100,7 @@ The value of the `$description` property MUST be a plain JSON string, for exampl
 {
```
[the remark this answers is not in the document]
Comment (c1rrus) [reply technical-reports/format/design-token.md:100]: It's the still correct that `$value` is the only required property on a design token object. `$type` can still be omitted, as long as one of the parent groups has `$type`, or it's a reference to another token.

File: technical-reports/format/design-token.md (hunk 2 above)
Comment (c1rrus) [reply technical-reports/format/design-token.md:16]: Agreed. I'll make that change in a moment.

File: technical-reports/format/design-token.md (hunk 3 above)
Comment (c1rrus) [reply technical-reports/format/design-token.md:128]: Good spot! I'll change it along the lines you've suggested.

Longer term, I think quite a few spec chapters could probably do with some re-organising. But that's for another day... ;-)

File: technical-reports/format/design-token.md [hunk 5]
```diff
@@ -113,7 +125,7 @@ The `$type` property can be set on different levels:
 - at the group level
 - at the token level
 
-The `$type` property MUST be a plain JSON string, whose value is `string`, `number`, `boolean`, `object`, `array`, `null` or one of the values specified in respective [type chapters](#types). The value of `$type` is case-sensitive.
+The `$type` property MUST be a plain JSON string, whose value is one of the values specified in this specification's respective type definitios. The value of `$type` is case-sensitive.
```
Comment (romainmenke) [inline technical-reports/format/design-token.md:128]:
```suggestion
The `$type` property MUST be a plain JSON string, whose value is one of the values specified in this specification's respective type definitions. The value of `$type` is case-sensitive.
```

typo :)

File: technical-reports/format/design-token.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinmpowell) [reply technical-reports/format/design-token.md:100]: @c1rrus I think it would be clearer to state that type is required along with name and value. Something like:

> Name, value and type are required. `$type` can be omitted from an individual token definition if defined on a parent group. See more information on Types below.

**Review by dbanksdesign — APPROVED** [review]
LGTM! Thank you for putting this together.

**Review by ilikescience — APPROVED** [review]
LGTM!

File: technical-reports/format/design-token.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (c1rrus) [reply technical-reports/format/design-token.md:100]: @kevinmpowell If it's ok with you, I'd like to leave this as is and merge this PR given that @dbanksdesign & @ilikescience both approved it now.

I think this is a symptom of a broader issue that would be best addressed in a separate PR: The spec's current use of the word "property" is overloaded. There are:

- _Syntax_ properties - i.e. the `$whatever` properties that are written in the files.
- _Model_ properties (someone please suggest a better name! 😛) - the properties that the (parsed) tokens and groups have.

Perhaps we need to agree on terminology to differentiate between these two meanings and then apply that consistently throughout the whole spec.

I'd argue that the for the current statement that "$value is the only required property for a token" is correct in the _syntax_ sense. However, in the _model_ sense, every token has at least name, value and type.
